grpc-circuitbreaker-utils/src/main/java/org/hypertrace/circuitbreaker/grpcutils/resilience/ResilienceCircuitBreakerInterceptor.java

```diff
@@ -0,0 +1,142 @@
+
+  public ResilienceCircuitBreakerInterceptor(
+      CircuitBreakerConfiguration<?> circuitBreakerConfiguration, Clock clock) {
+    this.circuitBreakerConfiguration = circuitBreakerConfiguration;
+    this.clock = clock;
+    this.resilienceCircuitBreakerConfigMap =
+        ResilienceCircuitBreakerConfigConverter.getCircuitBreakerConfigs(
+            circuitBreakerConfiguration.getCircuitBreakerThresholdsMap());
+    this.resilicenceCircuitBreakerRegistry =
+        new ResilienceCircuitBreakerRegistryProvider(
+                circuitBreakerConfiguration.getDefaultThresholds())
+            .getCircuitBreakerRegistry();
+    this.resilienceCircuitBreakerProvider =
+        new ResilienceCircuitBreakerProvider(
+            resilicenceCircuitBreakerRegistry, resilienceCircuitBreakerConfigMap);
+  }
+
+  public ResilienceCircuitBreakerInterceptor(
+      CircuitBreakerConfiguration<?> circuitBreakerConfiguration,
+      Clock clock,
+      CircuitBreakerRegistry resilicenceCircuitBreakerRegistry,
+      ResilienceCircuitBreakerProvider resilienceCircuitBreakerProvider) {
+    this.circuitBreakerConfiguration = circuitBreakerConfiguration;
+    this.resilienceCircuitBreakerConfigMap =
+        ResilienceCircuitBreakerConfigConverter.getCircuitBreakerConfigs(
+            circuitBreakerConfiguration.getCircuitBreakerThresholdsMap());
+    this.resilicenceCircuitBreakerRegistry = resilicenceCircuitBreakerRegistry;
+    this.resilienceCircuitBreakerProvider = resilienceCircuitBreakerProvider;
+    this.clock = clock;
+  }
+
+  @Override
+  protected boolean isCircuitBreakerEnabled() {
+    return circuitBreakerConfiguration.isEnabled();
+  }
+
+  @Override
+  protected <ReqT, RespT> ClientCall<ReqT, RespT> createInterceptedCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    return new ForwardingClientCall.SimpleForwardingClientCall<>(
+        next.newCall(method, callOptions)) {
+      CircuitBreaker circuitBreaker;
+      String circuitBreakerKey;
+
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        Instant startTime = clock.instant();
+        // Wrap response listener to track failures
+        Listener<RespT> wrappedListener =
+            wrapListenerWithCircuitBreaker(responseListener, startTime);
+        super.start(wrappedListener, headers);
+      }
+
+      @SuppressWarnings("unchecked")
+      @Override
+      public void sendMessage(ReqT message) {
+        CircuitBreakerConfiguration<ReqT> config =
+            (CircuitBreakerConfiguration<ReqT>) circuitBreakerConfiguration;
+        if (config.getKeyFunction() == null) {
+          log.debug("Circuit breaker will apply to all requests as keyFunction config is not set");
+          circuitBreakerKey = config.getDefaultCircuitBreakerKey();
+        } else {
+          circuitBreakerKey = config.getKeyFunction().apply(RequestContext.CURRENT.get(), message);
```

Comment:
We should support there being no default too - e.g. I can specify a default or not, overrides or not. If no key and no default, then it would not skip any circuit breaker.

Reply:
We may also want to check that the message matches the class if there is a class specified - either doing that here, or wrapping the config function inside the config works too. If we don't, we're requiring the caller to do it without type safety (because we're using the cast on line 89 to make the assumption they already did) and if they don't they'll get a runtime error.

Reply:
Even without defaults provided from consumer it will still take defaults and it will work.

Reply:
Added validation for Request class mismatch.